Create Geo Summary Files and update macros for new homeownership cost…

RAGarMir opened the pull request:
@ptatian I ran the summary files and kept getting errors re: the following macro which I also updated, Macros/ACS_summary_geo.sas.

lhendey commented:
@RAGarMir at least one error in ACS_summary_geo_source_tr_vars.sas = your new section of code (line 5380) shouldn't begin with %else %do. need an %if statement - like  `%if &_last_year. >= 2023  %then %do;  `  but @ptatian should take a look.

ptatian reviewed (changes requested):
@RAGarMir Please see my comments. One is an error that caused the program not to run. But I would also like you to make some changes to variable names and add the variable labels. Please let me know if you have any questions. CC: @lhendey

RAGarMir commented:
@ptatian I have updated the variable names and fixed the error you and Leah pointed out. I also added the variable label statements. I think there may be a couple that are too long, I have another version of the variable names without the 'num' prefix in case they have to be used. However, I batch submitted the update summary files and did not see any errors.

ptatian commented on Macros/ACS_summary_geo.sas:
```diff
@@ -1086,6 +1086,20 @@
+		IOM_Mort_CB_&_years. IOM_Mort_SCB_&_years. IOM_Mort_CBC_&_years. IOM_NoMort_CB_&_years. IOM_NoMort_SCB_&_years. IOM_NoMort_CBC_&_years. IOM_AllOwn_CB_&_years. IOM_AllOwn_SCB_&_years. IOM_AllOwn_CBC_&_years. IOM_Rent_CB_&_years. IOM_Rent_SCB_&_years. IOM_Rent_CBC_&_years. 
+		APIIOM_Mort_CB_&_years. APIIOM_Mort_SCB_&_years. APIIOM_Mort_CBC_&_years. APIIOM_NoMort_CB_&_years. APIIOM_NoMort_SCB_&_years. APIIOM_NoMort_CBC_&_years. APIIOM_AllOwn_CB_&_years. APIIOM_AllOwn_SCB_&_years. APIIOM_AllOwn_CBC_&_years. APIIOM_Rent_CB_&_years. APIIOM_Rent_SCB_&_years. APIIOM_Rent_CBC_&_years.
+
+	%else %if &_last_year. >=2017 %then %do;    
```
@RAGarMir There was another %else %if that you inserted without a corresponding %if before it. This one should also be changed to a %if. A semicolon is also missing at the end of the previous statement. `APIIOM_Rent_CBC_&_years. ;` These generated the following errors.

```
ERROR: There is no matching %IF statement for the %ELSE.
ERROR: A dummy macro will be compiled.
ERROR: Macro keyword IF appears as text.
ERROR: There is no matching %DO statement for the %END. This statement will be ignored.
```

I pushed these fixes, but will have to review the rest of the code later today.

ptatian commented:
@RAGarMir The code now runs with the changes I made. Our code produces listing output when a new data set is created. The code previously was not generating the listing output for the data sets because it was not running properly. That is another clue that things are not quite right. 

I pushed an additional change to eliminate some duplicate vars when the summary files are created. That is something that was introduced at an earlier stage. 

We still have at least a few problems to resolve, however. You are correct that some variable names are too long. That is indicated by these messages:

```
ERROR: The variable named NumnomortsvrecstbrdenAIOM_2019_23 contains more than 32 characters.
```

**You will have to reduce the variable names so that the total length (including the _2019_23 part) is 32 characters or less. **

Second problem is in the syntax of the %moe_sum() macro. In the LOG you will see messages like this:
```
MPRINT(ACS_SUMMARY_GEO_SOURCE_TR_VARS):   Nummortcstbrden_2019_23=
ERROR: All positional parameters must precede keyword parameters.
```
When you add this code, you left commas between the variable names. 
```
Nummortcstbrden_&_years.= %moe_sum( var=b25140bm3, b25140im3, b25140hm3, b25140cm3, b25140fm3, b25140gm3, b25140dm3, b25140em3);
```
That is how you do it with a sum() function. But this is a special macro that we wrote. In macros, commas are used to separate the parameters, but in this marco there is only one parameter, var=, and the list of vars that follows is the value assigned to that parameter. **The correct syntax for %moe_sum() is to leave out the commas.**
```
Nummortcstbrden_&_years.= %moe_sum( var=b25140bm3 b25140im3 b25140hm3 b25140cm3 b25140fm3 b25140gm3 b25140dm3 b25140em3);
```

Please make those two corrections and then recheck the code. Let's see if we can get this wrapped up! Thank you.

RAGarMir commented:
@ptatian I made the corrections and checked the code, I was not able to see any other issues.

ptatian commented:
Thanks, @RAGarMir. I will take a look.

ptatian commented on Macros/ACS_summary_geo_source_tr_vars.sas:
```diff
@@ -5363,6 +5366,293 @@
+				rentcstbrdenW_&_years.= b25140he11;
+				rentsvrecstbrdenW_&_years.= b25140he12;
+				rentcstbrdencalcW_&_years.= sum(b25140he11, b25140he12)
+
```
Missing semicolon at end of line above.

ptatian replied:
Generated error message: 
```
NOTE 138-205: Line generated by the macro variable "_YEARS".
111              mortcstbrdenH_2019_23
                 _____________________
                 22
ERROR 22-322: Syntax error, expecting one of the following: !, !!, &, *, **, +, -, /, ;, <, <=, <>, =, >, ><, >=, AND, EQ, 
              GE, GT, IN, LE, LT, MAX, MIN, NE, NG, NL, NOTIN, OR, ^=, |, ||, ~=.  
```

This is typical of a missing semicolon error message where the _next_ statement is the one that gets flagged because SAS thinks it is part of the previous statement.

ptatian commented on Macros/ACS_summary_geo_source_tr_vars.sas:
```diff
@@ -5363,6 +5366,293 @@
+				rentcstbrdenH_&_years.= b25140ie11;
+				rentsvrecstbrdenH_&_years.= b25140ie12;
+				rentcstbrdencalcH_&_years.= sum(b25140ie11, b25140ie12)
+
```
Missing semicolon at end of line above.

ptatian commented on Macros/ACS_summary_geo_source_tr_vars.sas:
```diff
@@ -5363,6 +5366,293 @@
+				rentcstbrdenB_&_years.= b25140be11;
+				rentsvrecstbrdenB_&_years.= b25140be12;
+				rentcstbrdencalcB_&_years.= sum(b25140be11, b25140be12)
+
```
Missing semicolon at end of line above.

ptatian commented on Macros/ACS_summary_geo_source_tr_vars.sas:
```diff
@@ -5363,6 +5366,293 @@
+				rentcstbrdenA_&_years.= sum(b25140de11, b25140ee11);
+				rentsvrecstbrdenA_&_years.= sum(b25140de12, b25140ee12);
+				rentcstbrdencalcA_&_years.= sum(b25140de11, b25140de12, b25140ee11, b25140ee12)
+
```
Missing semicolon at end of line above.

ptatian commented on Macros/ACS_summary_geo_source_tr_vars.sas:
```diff
@@ -5363,6 +5366,293 @@
+				mortcstbrdenW_&_years.= " Owners with Mortgage who are cost-burdened, Non-Hispanic White, &_years_dash "
+				mortsvrecstbrdenW_&_years.= " Owners with Mortgage who are severely-cost burdened, Non-Hispanic White, &_years_dash "
+				mortcstbrdencalcW_&_years.= " Owners with mortgage with cost burden calculated, Non-Hispanic White, &_years_dash " 
+				nomortcstbrdenW_&_years.= " Owners without Mortgage who are cost-burdened, Non-Hispanic White, &years_dash "
```
`&years_dash` should be `&_years_dash`

ptatian replied:
Generated this warning message:
```
WARNING: Apparent symbolic reference YEARS_DASH not resolved.
```

ptatian reviewed (changes requested):
@RAGarMir We have a few more problems to fix. Please see comments in review. Let me know if you have any questions.

RAGarMir commented:
Hi @ptatian I made the fixes to the code, added semi-colons where they were missing and made the '&years_dash' into '&_years_dash'.

ptatian commented:
@RAGarMir @lhendey In addition to what I flagged above, I checked the definition of the variables `mortcstbrden_&_years., mortsvrecstbrden_&_years., mortcstbrdencalc_&_years., ..., rentcstbrdencalc_&_years.`. These variables are supposed to be based on the _total_ owner and renter population, but currently they are not calculated correctly. For instance, 
```
mortcstbrden_&_years.= sum( b25140be3, b25140ce3, b25140de3, b25140ee3 b25140fe3, b25140ge3, b25140he3, b25140ie3 );
```
(I re-sorted the varables from the original.) This includes Black, Am. Indian, Asian, ..., White not Hisp/Latino, and Hispanic or Latino. But Hispanic or Latino overlaps with the other racial groups. The correct way to get the total is using only the race tables:  `b25140ae3, b25140be3, b25140ce3, b25140de3, b25140ee3 b25140fe3, b25140ge3`. That is, adding White alone and dropping White alone, not Hispanic or Latino and Hispanic or Latino. 

Please make this correction to this variables.

ptatian commented:
@RAGarMir The "calc" variables are also not correct. For instance, you have:
```
				mortcstbrdenW_&_years.= b25140he3;
				mortsvrecstbrdenW_&_years.= b25140he4;
				mortcstbrdencalcW_&_years.= sum(b25140he3, b25140he4);
```
b25140he3 is "Over 30 Percent" and b25140he4 is "Over 50 Percent," but the sum of those (b25140he3 + b25140he4) is not the total number of owner households with cost burdent calculated. The correct denominator for these cost burdens is `b25140he2 -b25140he5` ("Owned Units with a Mortgage" minus "Not Calculated"). 

This is going to be a problem for the MOE calculation for mortcstbrdencalcW_, etc. because the formulas don't do differences between cells. For the MOE for the "calc" vars, I would just use the MOE of the total of the group (in this case, "Owned Units with a Mortgage" or b25140he2) as a proxy.

ptatian commented:
I apologize for not flagging these problems sooner.

RAGarMir commented:
No problem @ptatian I am working on these changes, I just have quick formatting question. What would be the proper way to format the following correction:
rentcstbrdencalcA_&_years.= sum(b25140de11, b25140de12, b25140ee11, b25140ee12)

Is it:
rentcstbrdencalcA_&_years.= sum(b25140de10 - b25140de13, b25140ee10 - b25140ee13);

or is it:
rentcstbrdencalcA_&_years.= sum(b25140de10, b25140ee10 - b25140de13, b25140ee13);

Or are neither correct?

lhendey commented:
> @RAGarMir @lhendey In addition to what I flagged above, I checked the definition of the variables `mortcstbrden_&_years., mortsvrecstbrden_&_years., mortcstbrdencalc_&_years., ..., rentcstbrdencalc_&_years.`. These variables are supposed to be based on the _total_ owner and renter population, but currently they are not calculated correctly. For instance,
> 
> ```
> mortcstbrden_&_years.= sum( b25140be3, b25140ce3, b25140de3, b25140ee3 b25140fe3, b25140ge3, b25140he3, b25140ie3 );
> ```
> 
> (I re-sorted the varables from the original.) This includes Black, Am. Indian, Asian, ..., White not Hisp/Latino, and Hispanic or Latino. But Hispanic or Latino overlaps with the other racial groups. The correct way to get the total is using only the race tables: `b25140ae3, b25140be3, b25140ce3, b25140de3, b25140ee3 b25140fe3, b25140ge3`. That is, adding "White alone" and dropping "White alone, not Hispanic or Latino" and "Hispanic or Latino."
> 
> Please make this correction to these variables.

@ptatian - shouldn't we just use table B25140 to calculate the totals? (vs. adding race tables?)

ptatian commented:
@lhendey I checked that, but we didn't pull table B25140.

ptatian commented:
@RAGarMir You need to subtract "not calculated" cell from "total" cell for each racial group that you are including in the variable. So the first expression you had is correct:

rentcstbrdencalcA_&_years.= sum(b25140de10 - b25140de13, b25140ee10 - b25140ee13);

RAGarMir commented:
@ptatian I made the requested changes

RAGarMir commented:
@ptatian I made the updates to the moe variable names. I ran the summary files and saw no errors.

ptatian commented:
Thank you, @RAGarMir. I will review.

ptatian commented:
@RAGarMir I pushed a couple of commits just to clean up some redundant variable names in the summarizing portion of the code. I also renamed the variable MOwnerCostsPctHHInc to MedOwnerCostsPctHHInc. This is not one of your vars, but was created earlier. Having just "M" at the front of a variable name is how we designate MOE vars, so I renamed to be more consistent with other median variables that start with "Med."

The code now runs without any errors, but there is still a problem. The last step in checking code is to look at the _results_ that the code produces. In this case, those are the new data sets with the summary variables. We generate Proc Contents and Proc Means output so that we can examine those results. When you look at the files that the program is producing, do you notice anything that isn't right?

ptatian reviewed (approved):
@RAGarMir These programs are now ready for the final batch submit.
